### ISSUE TITLE
OS collector fails with message "failed collecting os metrics:<nil> CreateFile D:\pagefile.sys: The system cannot find the file specified."

### ISSUE THREAD

**Issue by chavalikesavanand** [prometheus-community/windows_exporter#1069]
We are using 0.19.0 version with Prometheus 2.34.0.
windows_exporter_collector_success shows 0 for os collector. All other collectors succeeded.

we checked that pagefile is indeed enabled only on d: drive with "system managed" size. It is disabled on c: drive

we already tried executing lodctr /R many times and restarting windows exporter windows service many times.
Windows exporter windows service runs in local system account.

Windows exporter config:

collectors:
  enabled: cpu,cpu_info,cs,iis,logical_disk,logon,memory,net,os,process,service,system,time

![image](https://user-images.githubusercontent.com/93990393/192543483-1205d5ed-dbb4-4a56-947d-76fce9507dc1.png)

**Comment by breed808** [comment]
Strange, this should have been resolved in #954. Are you seeing any relevant output from the exporter for the `os` collector?

**Comment by chavalikesavanand** [comment]
thanks @breed808 for quick turn around. Appreciate it.
Basically, it logs this error in windows event logs (we configured windows event logs to be log destination for exporter) and quits OS collector. I found the following lines in OS Collector code. It is quitting OS metrics collection if page file read gives error. Is it possible to continue with other OS metrics? Also i think it is better to control this error log, may be log only once or once in some number of scrapes etc? It is cluttering our event logs...

ln 182 to 189

pagingFiles, _, pagingErr := memManKey.GetStringsValue("ExistingPageFiles")
	// Get build number and product name from registry
	ntKey, err := registry.OpenKey(registry.LOCAL_MACHINE, `SOFTWARE\Microsoft\Windows NT\CurrentVersion`, registry.QUERY_VALUE)
	defer ntKey.Close()

	if err != nil {
		return nil, err
	}

Is it possible to continue above without quitting?

**Comment by breed808** [comment]
I've tried to reproduce this on a spare Windows 10 system but have failed. The exporter continues without failure, with paging on D:\ set to "system managed" and paging on C:\ disabled.

![Screenshot_20221005_201206](https://user-images.githubusercontent.com/3103817/194036897-9bfc237d-5aa3-4c30-8c03-59991dc01a53.png)

The only meaningful difference I can see is the addition of the `PagingFiles` key, but the absence of this shouldn't cause the `os` collector to fail.

**Comment by ReneZeidler** [comment]
I'm running into the same/related issue.

The OS collector fails with the error: `failed collecting os metrics:<nil> CreateFile B:\pagefile.sys: The system cannot find the file specified.`

The registry value for `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Session Manager\Memory Management\ExistingPageFiles` is:

    \??\B:\pagefile.sys
    \??\C:\pagefile.sys

The system is indeed configured to create a page file on B: with system managed size:
![image](https://user-images.githubusercontent.com/5558742/206697408-9b06b75b-a59e-43f2-b0a9-b14aa6973b43.png)

However, the error message is correct, and the page file doesn't actually exist on B:. The page file on C: exists with the correct size. Windows itself doesn't seem to care that the file is missing, it just doesn't show up in the Performance Monitor:
![image](https://user-images.githubusercontent.com/5558742/206702304-b80c4365-cf40-43b4-9d2c-841bd72ac5b9.png)
It reports the total page file size as 400 MB, which is just the size of the C: pagefile.

I don't know why the file doesn't exist, I couldn't find any warnings or errors in the event viewer. The disk is writable but completely empty. Maybe Windows just decided that 0 MB is the correct size and just doesn't create the file, who knows.

This whole situation seems a bit nonsensical and misconfigured, but it is apparently a configuration that Windows allows, and it is the one I have to deal with for now.

---

My issue is that this missing file causes the whole OS collector to fail.

I think an appropriate solution would be to just skip a page file if it is missing and don't add anything to the total, instead of returning an error here:
https://github.com/prometheus-community/windows_exporter/blob/master/collector/os.go#L205-L208

**Comment by breed808** [comment]
Great work investigating that :+1:. Float values in golang initialise at 0 so omitting the error check is fine here.

@ReneZeidler could I get you to test the feature branch in #1108 before I merge it?

**Comment by ReneZeidler** [comment]
@breed808 

Thanks for the fix. Unfortunately, the exporter now crashes:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x38 pc=0x14d239b]

goroutine 40 [running]:
github.com/prometheus-community/windows_exporter/collector.(*OSCollector).collect(0xc0000faa80, 0xc0000c65b8, 0x0?)
        /home/r718037/promu/windows_exporter/collector/os.go:209 +0x163b
github.com/prometheus-community/windows_exporter/collector.(*OSCollector).Collect(0x0?, 0x0?, 0x0?)
        /home/r718037/promu/windows_exporter/collector/os.go:136 +0x1d
main.execute({0x174ff2a, 0x2}, {0x185cfe0, 0xc0000faa80}, 0x0?, 0x0?)
        /home/r718037/promu/windows_exporter/exporter.go:200 +0x84
main.windowsCollector.Collect.func2({0x174ff2a, 0x2}, {0x185cfe0?, 0xc0000faa80?})
        /home/r718037/promu/windows_exporter/exporter.go:141 +0x9b
created by main.windowsCollector.Collect
        /home/r718037/promu/windows_exporter/exporter.go:139 +0x5c5
```

The [line](https://github.com/prometheus-community/windows_exporter/blob/fb062dc02056958746f44ee7e0cf07bce81ff56a/collector/os.go#L209)
```go
fsipf += float64(file.Size())
```
needs to be wrapped in an `else` block. If I do that, everything works as expected and the OS exporter runs without error.

**Comment by chavalikesavanand** [comment]
When this fix will be released?

**Comment by breed808** [comment]
[Release candidate](https://github.com/prometheus-community/windows_exporter/releases/tag/v0.21.0-rc1) for v0.21.0 has been made available. v0.21.0 release will be available in about a week if no blocking issues are raised.

**Comment by chavalikesavanand** [comment]
Any update on the release please? we are waiting for this fix

**Comment by chavalikesavanand** [comment]
@breed808 any updates on the release please?

**Comment by breed808** [comment]
@chavalikesavanand  did the [v0.21.0 release candidate](https://github.com/prometheus-community/windows_exporter/releases/tag/v0.21.0-rc1) resolve your issue?

I'm hoping to have the official v0.21.0 released after merging #1133; unfortunately life has got in the way and delayed me a few weeks.

**Comment by chavalikesavanand** [comment]
@breed808 any update on release .21 please?
